### PR TITLE
[SPARK-39112][SQL] UnsupportedOperationException if spark.sql.ui.explainMode is set to cost

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add a new leaf like node `LeafNodeWithoutStats` and apply to the list:
- ResolvedDBObjectName
- ResolvedNamespace
- ResolvedTable
- ResolvedView
- ResolvedNonPersistentFunc
- ResolvedPersistentFunc

### Why are the changes needed?

We enable v2 command at 3.3.0 branch by default `spark.sql.legacy.useV1Command`. However this is a behavior change between v1 and c2 command.

- v1 command:
  We resolve logical plan to command at analyzer phase by `ResolveSessionCatalog`

- v2 commnd:
  We resolve logical plan to v2 command at physical phase by `DataSourceV2Strategy`

Foe cost explain mode, we will call `LogicalPlanStats.stats` using optimized plan so there is a gap between v1 and v2 command.
Unfortunately, the logical plan of v2 command contains the `LeafNode` which does not override the `computeStats`. As a result, there is a error running such sql:
```sql
set spark.sql.ui.explainMode=cost;
show tables;
```

```
java.lang.UnsupportedOperationException:
	at org.apache.spark.sql.catalyst.plans.logical.LeafNode.computeStats(LogicalPlan.scala:171)
	at org.apache.spark.sql.catalyst.plans.logical.LeafNode.computeStats$(LogicalPlan.scala:171)
	at org.apache.spark.sql.catalyst.analysis.ResolvedNamespace.computeStats(v2ResolutionPlans.scala:155)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.SizeInBytesOnlyStatsPlanVisitor$.default(SizeInBytesOnlyStatsPlanVisitor.scala:55)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.SizeInBytesOnlyStatsPlanVisitor$.default(SizeInBytesOnlyStatsPlanVisitor.scala:27)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlanVisitor.visit(LogicalPlanVisitor.scala:49)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlanVisitor.visit$(LogicalPlanVisitor.scala:25)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.SizeInBytesOnlyStatsPlanVisitor$.visit(SizeInBytesOnlyStatsPlanVisitor.scala:27)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.LogicalPlanStats.$anonfun$stats$1(LogicalPlanStats.scala:37)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.LogicalPlanStats.stats(LogicalPlanStats.scala:33)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.LogicalPlanStats.stats$(LogicalPlanStats.scala:33)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.stats(LogicalPlan.scala:30)
```

### Does this PR introduce _any_ user-facing change?

yes, bug fix

### How was this patch tested?

add test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @HyukjinKwon @wangyum @MaxGekk

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/v2ResolutionPlans.scala [hunk 1]
```diff
@@ -140,11 +141,19 @@ case class UnresolvedDBObjectName(nameParts: Seq[String], isNamespace: Boolean)
   override def output: Seq[Attribute] = Nil
 }
 
+/**
+ * A resolved leaf like node that its statistics is no meaning.
+ */
+trait ResolvedLeafObject extends LogicalPlan with LeafLike[LogicalPlan] {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/v2ResolutionPlans.scala:147]: nit: `LeafNodeWithoutStats` seems better

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 2]
```diff
@@ -528,6 +528,23 @@ class ExplainSuite extends ExplainSuiteHelper with DisableAdaptiveExecutionSuite
         "== Analyzed Logical Plan ==\nCreateViewCommand")
     }
   }
+
+  test("SPARK-39112: UnsupportedOperationException if spark.sql.ui.explainMode is set to cost") {
+    withSQLConf(SQLConf.UI_EXPLAIN_MODE.key -> "cost") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:533]: can we simply test with `EXPLAIN COST ...` command?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/v2ResolutionPlans.scala [hunk 3]
```diff
@@ -140,11 +141,19 @@ case class UnresolvedDBObjectName(nameParts: Seq[String], isNamespace: Boolean)
   override def output: Seq[Attribute] = Nil
 }
 
+/**
+ * A resolved leaf like node that its statistics is no meaning.
+ */
+trait LeafNodeWithoutStats extends LogicalPlan with LeafLike[LogicalPlan] {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/v2ResolutionPlans.scala:147]: Can we make it an abstract class and extends `LeafNode`? I'm a bit worried about changing the class hierarchy unnecessarily in this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/v2ResolutionPlans.scala [hunk 4]
```diff
@@ -140,11 +141,19 @@ case class UnresolvedDBObjectName(nameParts: Seq[String], isNamespace: Boolean)
   override def output: Seq[Attribute] = Nil
 }
 
+/**
+ * A resolved leaf like node that its statistics is no meaning.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/v2ResolutionPlans.scala:145]:
```suggestion
 * A resolved leaf like node that its statistics has no meaning.
```

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 5]
```diff
@@ -528,6 +528,21 @@ class ExplainSuite extends ExplainSuiteHelper with DisableAdaptiveExecutionSuite
         "== Analyzed Logical Plan ==\nCreateViewCommand")
     }
   }
+
+  test("SPARK-39112: UnsupportedOperationException if spark.sql.ui.explainMode is set to cost") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:532]: the test name needs update

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/v2ResolutionPlans.scala (hunk 3 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/v2ResolutionPlans.scala]: changed to `trait LeafNodeWithoutStats extend LeafNode`

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Pull request closed without merging**